src/dynamics/joint.rs

```diff
@@ -1,9 +1,73 @@
-use crate::dynamics::GenericJoint;
 use bevy::prelude::*;
 use rapier::dynamics::{ImpulseJointHandle, MultibodyJointHandle};
 
 pub use rapier::dynamics::{JointAxesMask, JointAxis, MotorModel};
 
+use super::{FixedJoint, GenericJoint, PrismaticJoint, RevoluteJoint, RopeJoint, SpringJoint};
+
+#[cfg(feature = "dim3")]
+use super::SphericalJoint;
+
+/// Wrapper enum over a specific joint.
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum JointDescription {
+    /// See [`FixedJoint`]
+    FixedJoint(FixedJoint),
+    /// See [`GenericJoint`]
+    GenericJoint(GenericJoint),
+    /// See [`PrismaticJoint`]
+    PrismaticJoint(PrismaticJoint),
+    /// See [`RevoluteJoint`]
+    RevoluteJoint(RevoluteJoint),
+    /// See [`RopeJoint`]
+    RopeJoint(RopeJoint),
+    /// See [`SphericalJoint`]
+    #[cfg(feature = "dim3")]
+    SphericalJoint(SphericalJoint),
+    /// See [`SpringJoint`]
+    SpringJoint(SpringJoint),
+}
+
+impl JointDescription {
+    /// The underlying generic joint.
+    pub fn generic_joint(&self) -> &GenericJoint {
+        match self {
+            JointDescription::FixedJoint(j) => &j.data,
+            JointDescription::GenericJoint(j) => j,
+            JointDescription::PrismaticJoint(j) => &j.data,
+            JointDescription::RevoluteJoint(j) => j.data(),
+            JointDescription::RopeJoint(j) => j.data(),
+            #[cfg(feature = "dim3")]
+            JointDescription::SphericalJoint(j) => j.data(),
+            JointDescription::SpringJoint(j) => j.data(),
+        }
+    }
+    /// The underlying generic joint.
+    pub fn generic_joint_mut(&mut self) -> &mut GenericJoint {
+        match self {
+            JointDescription::FixedJoint(ref mut j) => &mut j.data,
+            JointDescription::GenericJoint(ref mut j) => j,
+            JointDescription::PrismaticJoint(ref mut j) => &mut j.data,
+            JointDescription::RevoluteJoint(ref mut j) => &mut j.data,
+            JointDescription::RopeJoint(ref mut j) => &mut j.data,
+            #[cfg(feature = "dim3")]
+            JointDescription::SphericalJoint(ref mut j) => &mut j.data,
+            JointDescription::SpringJoint(ref mut j) => &mut j.data,
+        }
+    }
+}
```

Comment:
These could be `AsRef`/`AsMut` implementations instead.

Reply:
true, I'll do that.

For the story behind this explicit naming: ideally I would have liked to wrap our joints (`bevy_rapier::FixedJoint`, etc...) around the correct joint from rapier (`rapier::FixedJoint`, etc.), so a user could use the relevant functions from rapier if some functions were not exposed by bevy_rapier (it can easily be forgotten...), but I had difficulties, probably the builder had to be updated significantly so I dropped the idea.

See https://github.com/dimforge/bevy_rapier/pull/529 for similar reasoning logic.